app/code/community/Bolt/Boltpay/Helper/Bugsnag.php

```diff
@@ -154,19 +154,14 @@ public function setBoltTraceId($traceId)
         $this->boltTraceId = $traceId;
     }
 
+    /**
+     * The Bolt TraceID set in the header file
+     *
+     * @return string|null  it will be a string if it is set, otherwise null
+     */
     public function getBoltTraceId() 
     {
-        if(isset($this->boltTraceId)) {
-            return $this->boltTraceId;
-        }
-
-        $traceId = @$_SERVER['HTTP_X_BOLT_TRACE_ID'];
-
-        if(!empty($traceId)) {
-            return $traceId;
-        }
-
-        return null;
+        return $this->boltTraceId ?: @$_SERVER['HTTP_X_BOLT_TRACE_ID'];
```

Comment:
I sugget to move $_SERVER to the separate method and then we can use something like that:

Create di.xml and
```xml
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <!-- Restore original visitor IP from CloudFlare header -->
    <type name="Magento\Framework\HTTP\PhpEnvironment\RemoteAddress">
        <arguments>
            <argument name="alternativeHeaders" xsi:type="array">
                <item name="http_cf_connecting_ip" xsi:type="string">HTTP_CF_CONNECTING_IP</item>
            </argument>
        </arguments>
    </type>
</config>
```

and then we can use something like:
```php
    /**
     * @var \Magento\Framework\HTTP\PhpEnvironment\RemoteAddress
     */
    protected $_remoteAddressInstance;

...

    /**
     * Used to get remote address, in which cookie data is set
     *
     * @return string returns remote address
     */
    public function getRemoteAddress()
    {
        return $this->_remoteAddressInstance->getRemoteAddress();
    }
```

Reply:
@YevhenShBolt  Why would I make something that is one line of code into something that is four lines of code spanning across multiple methods?

To me, that is highly unnecessary and makes the code harder to read and needlessly more complex.

I liken your suggestion to `goto` construct, that was used in the late 50s, 60s, 70s, and 80s.  In that style, we are jumping around in the code for no good reason.  

If there was some sort of sanitizing of values that were needed in looking into the array, then there is a point.  But, in this case, there seems to be no point, and we do not need to wrap a method around an array to pull its raw values.

My point is this:
Look at the old code, 8 lines.
Look at the new code, equivalent and only 1 line.

And what I understand the suggestion to be is  something like this:

```
...
return $this->boltTraceId ?: $this->_getHeaderField('HTTP_X_BOLT_TRACE_ID');
...

/**
 * Method to an element from the $_SERVER array
 */
 private function _getHeaderField( $key ) {
     return @$_SERVER[$key];
}
```

So, one line now becomes six with a fairly useless and logically obscuring method.

I disagree with this approach and consider it an anti-pattern.

Reply:
My goal is to simply it.
I want to make something like:

```php
...
return $this->boltTraceId ?: $this->_getServerBoltTraceId();
...

/**
 * Method to get an element from the $_SERVER array
 */
 private function _getServerBoltTraceId() {
     return $_SERVER['HTTP_X_BOLT_TRACE_ID'];
}
```

Here, we can modify it as we want and will not touch the base view of `public function getBoltTraceId();`
For example, if the merchant will use the Cloudflare or varnish the headers will be removed and we should use another name or logic to determine which header we needed.
In such way, i see that such method will be useful.

@daisy1754  and @mccottry . Guys, if you think that for now, it will be fairly useless - please ignore my recommendation. Maybe I'm getting paranoid ;-)

Reply:
@YevhenShBolt IMO, we should stop adding methods like this to the code for the reasons I explained above.  In your example with Cloudflare and Varnish, if it were on a merchant's site and this sort of problem arose, then you still would have to adjust the method `public function getBoltTraceId();` because the private method could not be overridden.

Obviously, you could make your helper method of the helper method protected or public.  But, just re-read that... we have a one-line helper method for another one-line helper method that create 6 to 11 lines of code for something that can effectively be accomplished in one simple line.

When making a helper method, we should ask ourselves at least three fundamental questions.

1.)  Does this method significantly reduce the complexity of a code block?
2.)  Is this method likely to be reused in other than one place in the code?
3.)  Is it likely that I will need to override this specific functionality?

If you can make a strong case for any of these things, then a helper method is warranted.  Here, I think the method you will want to override would be `public function getBoltTraceId()`.  

For the cases you described above, (i.e. headers being rewritten), then I think the scope of these functions is too narrow for the projected problem and I would recommend handling that in another branch/pull request.

Reply:
ok